File: app/src/main/java/com/example/epamtraining/MainActivity.java
package com.example.epamtraining;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;

import java.util.Random;

public class MainActivity extends AppCompatActivity {

    private static final String TAG = "MainActivity";

    private DrawerLayout drawerLayout;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initToolbar();

        initDrawer();

    }

    public void initToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionbar = getSupportActionBar();
        if (actionbar != null) {
            actionbar.setDisplayHomeAsUpEnabled(true);
            actionbar.setHomeAsUpIndicator(R.drawable.ic_menu);
        }
    }

    public void initDrawer() {
        drawerLayout = findViewById(R.id.drawer_layout);
        NavigationView navigationView = findViewById(R.id.nav_view);

        final HeaderView headerView = navigationView
                .getHeaderView(0)
                .findViewById(R.id.header_view);

        headerView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                String color = generateGolor();
                headerView.updateImage(color);
                Log.i(TAG, "color: " + color);
            }
        });

        navigationView.setNavigationItemSelectedListener(
                new NavigationView.OnNavigationItemSelectedListener() {
                    @Override
                    public boolean onNavigationItemSelected(@NonNull MenuItem menuItem) {
                        menuItem.setChecked(true);
                        drawerLayout.closeDrawers();
                        selectDrawerItem(menuItem);

                        return true;
                    }
                });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                drawerLayout.openDrawer(GravityCompat.START);
                return true;
        }

        return super.onOptionsItemSelected(item);
    }

    public void selectDrawerItem(MenuItem menuItem) {
        Fragment fragment;
        Class fragmentClass;

        switch (menuItem.getItemId()) {
            case R.id.first_fragment_item:
                fragmentClass = FirstFragment.class;

                break;
            case R.id.second_fragment_item:
                fragmentClass = SecondFragment.class;

                break;
            default:
                fragmentClass = FirstFragment.class;

                break;
        }

        try {
            fragment = (Fragment) fragmentClass.newInstance();
            getSupportFragmentManager()
                    .beginTransaction()
                    .replace(R.id.content_frame, fragment).commit();
        } catch (Exception e) {
            Log.e(TAG, "selectDrawerItem: " + e);
        }
    }

    private String generateGolor() {
        Random random = new Random();
        int colorCode = random.nextInt(0xffffff + 1);
        String color = String.format("#%06x", colorCode);

        return color;
    }
}

File: app/src/main/java/com/example/epamtraining/activities/MainActivity.kt
package com.example.epamtraining.activities

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.ActionBar
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.GravityCompat
import com.example.epamtraining.R
import com.example.epamtraining.activities.LoginActivity.Companion.startAuth
import com.example.epamtraining.contracts.MainContract
import com.example.epamtraining.customViews.HeaderView
import com.example.epamtraining.models.User
import com.example.epamtraining.network.FirebaseAuth
import com.example.epamtraining.network.FirebaseDatabase
import com.example.epamtraining.presenters.MainPresenter
import com.example.epamtraining.repositories.MainRepository
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.toolbar.*
import kotlin.concurrent.thread


class MainActivity : AppCompatActivity(), MainContract.View {

    private val mainRepository = MainRepository()
    private val mainPresenter = MainPresenter(this, supportFragmentManager, this, mainRepository)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initToolbar()

        initDrawer()

        mainPresenter.checkCurrentUserToken()
    }

    override fun initHeaderView() {
        val headerView = mainNavigationView
                .getHeaderView(0)
                .findViewById<HeaderView>(R.id.main_header_view)

        thread {
            val user: User? = FirebaseDatabase.getUserInfo()
            runOnUiThread {
                headerView.initViews(user?.name.toString())
            }
        }
    }

    override fun initProfile() {
        mainNavigationView.setCheckedItem(R.id.drawerProfile)
        mainPresenter.selectDrawerItem(mainNavigationView.menu.getItem(0))
    }

    override fun initDrawer() {
        mainNavigationView.setNavigationItemSelectedListener { menuItem ->
            menuItem.isChecked = true
            drawerLayout.closeDrawers()
            mainPresenter.selectDrawerItem(menuItem)
            true
        }
    }

    override fun detachMainActivity() {
        finish()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                drawerLayout.openDrawer(GravityCompat.START)

                return true
            }
        }

        return super.onOptionsItemSelected(item)
    }

    override fun onBackPressed() {
        if (drawerLayout.isDrawerOpen(GravityCompat.START)) drawerLayout.closeDrawer(GravityCompat.START) else drawerLayout.openDrawer(GravityCompat.START)
    }

    override fun logout() {
        removeAllFragments()
        finish()
        FirebaseAuth.signOut()
        startAuth(this@MainActivity)
    }

    override fun initToolbar() {
        setSupportActionBar(mainToolbar)
        val actionbar: ActionBar? = supportActionBar

        actionbar?.apply {
            setDisplayHomeAsUpEnabled(true)
            setHomeAsUpIndicator(R.drawable.ic_menu)
        }
    }

    override fun removeAllFragments() {
        val fragmentManager = supportFragmentManager

        while (fragmentManager.backStackEntryCount > 0) {
            fragmentManager.popBackStackImmediate()
        }
    }

    companion object {
        fun startMainActivity(packageContext: Context?) {
            val intent = Intent(packageContext, MainActivity::class.java)
            packageContext?.startActivity(intent)
        }
    }
}